### PR TITLE
[Calyx] Replace calyx.program with module attribute calyx.entrypoint

### PR DESCRIPTION
**Pull request by andrewb1999**
Based on discussion about removing calyx.program in #3565. Allows the calyx dialect to integrate better with other dialects.

### CODE REVIEW

**Comment by andrewb1999** [comment]
CC: @mortbopet @mikeurbach @cgyurgyik @rachitnigam

**Review by cgyurgyik — APPROVED** [review]
Thanks for undertaking this! Approved when the continuous integrations pass.

**Comment by andrewb1999** [comment]
@cgyurgyik Looks like this passed CI, good to merge?

**Comment by cgyurgyik** [comment]
Yes, would you like to merge it yourself?

If so, you'll need to request commit access (below). If not, I can merge it for you.

https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h [hunk 1]
```diff
@@ -525,7 +523,7 @@ class FuncOpPartialLoweringPattern
   }
 
   /// Return the program lowering state for this pattern.
-  ProgramLoweringState &programState() const;
+  CalyxLoweringState &programState() const;
```
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:526]: stale comment, function name

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h [hunk 2]
```diff
@@ -539,7 +537,7 @@ class FuncOpPartialLoweringPattern
 private:
   mutable ComponentOp componentOp;
   mutable ComponentLoweringStateInterface *componentLoweringState = nullptr;
-  ProgramLoweringState &programLoweringState;
+  CalyxLoweringState &programLoweringState;
```
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:540]: stale variable name

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h [hunk 3]
```diff
@@ -607,7 +605,7 @@ class InlineCombGroups
                           calyx::GroupInterface originGroup,
                           calyx::GroupInterface recGroup, bool doInline) const;
 
-  calyx::ProgramLoweringState &pls;
+  calyx::CalyxLoweringState &pls;
```
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:608]: stale variable name

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h [hunk 4]
```diff
@@ -617,14 +615,14 @@ class RewriteMemoryAccesses
     : public calyx::PartialLoweringPattern<calyx::AssignOp> {
 public:
   RewriteMemoryAccesses(MLIRContext *context, LogicalResult &resRef,
-                        calyx::ProgramLoweringState &pls)
+                        calyx::CalyxLoweringState &pls)
```
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:618]: stale variable name

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h [hunk 5]
```diff
@@ -617,14 +615,14 @@ class RewriteMemoryAccesses
...
       : PartialLoweringPattern(context, resRef), pls(pls) {}
 
   LogicalResult partiallyLower(calyx::AssignOp assignOp,
                                PatternRewriter &rewriter) const override;
 
 private:
-  calyx::ProgramLoweringState &pls;
+  calyx::CalyxLoweringState &pls;
```
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:625]: stale variable name

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 6]
```diff
@@ -1230,7 +1219,7 @@ static LogicalResult verifyInstanceOpType(InstanceOp instance,
 
 LogicalResult InstanceOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
   Operation *op = *this;
-  auto program = op->getParentOfType<ProgramOp>();
+  auto program = op->getParentOfType<ModuleOp>();
```
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:1222]: stale variable name

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 7]
```diff
@@ -1187,7 +1175,7 @@ void AssignOp::print(OpAsmPrinter &p) {
 /// Lookup the component for the symbol. This returns null on
 /// invalid IR.
 ComponentOp InstanceOp::getReferencedComponent() {
-  auto program = (*this)->getParentOfType<ProgramOp>();
+  auto program = (*this)->getParentOfType<ModuleOp>();
```
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:1178]: stale variable name

File: lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp [hunk 8]
```diff
@@ -348,38 +348,22 @@ std::string ProgramLoweringState::blockName(Block *b) {
...
-    rewriter.setInsertionPointToStart(moduleBlock);
-    rewriter.insert(programOp);
-    *programOpOutput = programOp;
-  });
+  moduleOp->setAttr("calyx.entrypoint",
+                    rewriter.getStringAttr(topLevelFunction));
+  // rewriter.updateRootInPlace(moduleOp, [&] {
+  //   // Create ProgramOp
+  //   rewriter.setInsertionPointAfter(moduleOp);
+  // });
```
Comment (mortbopet) [inline lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp:366]: is this needed?

File: lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp [hunk 9]
```diff
@@ -415,7 +399,7 @@ calyx::ComponentOp FuncOpPartialLoweringPattern::getComponent() const {
   return componentOp;
 }
 
-ProgramLoweringState &FuncOpPartialLoweringPattern::programState() const {
+CalyxLoweringState &FuncOpPartialLoweringPattern::programState() const {
```
Comment (mortbopet) [inline lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp:402]: stale function name

File: test/Conversion/CalyxToFSM/lower-seq.mlir [hunk 10]
```diff
@@ -31,7 +31,7 @@
 // CHECK-NEXT: }
       
 
-calyx.program "main" {
+module attributes {calyx.entrypoint = "main"} {
```
Comment (mortbopet) [inline test/Conversion/CalyxToFSM/lower-seq.mlir:34]: These `calyx.entrypoint` attributes shouldn't be necessary for the `CalyxToFSM` passes to succeed.

**Review by mortbopet — COMMENTED** [review]
Some nits.

File: test/Conversion/CalyxToFSM/lower-seq.mlir (hunk 10 above)
Comment (andrewb1999) [reply test/Conversion/CalyxToFSM/lower-seq.mlir:34]: Yeah it's not required as far as I know. Should I leave them out of the test cases?

File: lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp (hunk 8 above)
Comment (andrewb1999) [reply lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp]: No, removed.

**Comment by andrewb1999** [comment]
@mortbopet look good now?

File: test/Conversion/CalyxToFSM/lower-seq.mlir (hunk 10 above)
Comment (mortbopet) [reply test/Conversion/CalyxToFSM/lower-seq.mlir:34]: I'd do that - tests should be as stripped as possible.

**Review by mortbopet — APPROVED** [review]
LGTM - Thanks!

One thing that we should consider for follow-up work is if we even _want_ a `calyx.entrypoint` attribute - I do not think that this should be an inherent property of the IR. As far as I can tell, that entry point only really makes sense for simulators, and not for the compiler itself.  
IMO, tools that require an explicit entry point should:
1. topologically sort the instance hierarchy to infer the top-level component _or_
2. explicitly ask the user about the entry point

A path to this is to implement `HWModuleLike, HWInstanceLike` interfaces for the Calyx ops, and use the [InstanceGraphBase.h](https://github.com/llvm/circt/blob/main/include/circt/Dialect/HW/InstanceGraphBase.h).
(looking at the code, I think that that graph should contain a `getInferredTopModule` operation which performs the topological sort...).

**Pull request merged**